store/src/java/com/zimbra/cs/filter/SieveToSoap.java

```diff
@@ -81,26 +81,50 @@ protected void visitRule(Node ruleNode, VisitPhase phase, RuleProperties props)
                 currentNestedRule.setChild(nestedRule);
             } else {               // first nested rule
                 // set it as child of root rule
+                if (null == currentRule) {
+                    initCurrentRule(props);
+                }
                 currentRule.setChild(nestedRule);
             }
             currentNestedRule = nestedRule;
         }
     }
 
+    private void initCurrentRule(RuleProperties props) {
+        currentRule = new FilterRule(getCurrentRuleName(), props.isEnabled);
+        currentRule.setFilterTests(new FilterTests(props.condition.toString()));
+        rules.add(currentRule);
+        currentRuleIndex++;
+        // When start working on the root rule, initialise the pointer to nested rule
+        currentNestedRule = null;
+    }
+
     @Override
     protected void visitIfControl(Node ruleNode, VisitPhase phase, RuleProperties props) {
         if (phase == VisitPhase.end) {
             numOfIfProcessingDone++; // number of if for which process is done
+            if (nunOfIfProcessingStarted == numOfIfProcessingDone) {
+                // If all the nested 'if' is completed, clean up the rule objects
+                currentRule = null;
+                currentNestedRule = null;
+            }
+            visitRule(ruleNode, VisitPhase.end, props);
             return;
         }
         nunOfIfProcessingStarted++;   // number of if for which process is started.
+        visitRule(ruleNode, VisitPhase.begin, props);
     }
 
     private boolean isNestedRule(){
         // in non nested case, only one process is started but not done.
         if(nunOfIfProcessingStarted == numOfIfProcessingDone+1){
             return false;
         }
+        // if an action is specified outside of the 'if' clause,
+        // the number of Started becomes equal to that of Done.
+        if (nunOfIfProcessingStarted == numOfIfProcessingDone) {
```

Comment:
In this case, this looks intuitionally readable 'if' block because the length of the block is short (one line), there is the similar condition before it, and the source code comment is added.